client-rpc/src/rpc/staking_rpc.rs

```diff
@@ -96,31 +107,150 @@ where
+                    format!("Unable to deserialize to_staking_address ({})", to_address),
+                )
+            })
+            .map_err(to_rpc_error)?;
         let attr = StakedStateOpAttributes::new(self.network_id);
+        let fee = self
+            .ops_client
+            .calculate_deposit_fee()
+            .map_err(to_rpc_error)?;
+        let total_amount = (amount + fee).map_err(to_rpc_error)?;
+        // 1. build a transfer transaction to make a UTXO which amount is `deposit_amount + fee`
+        let to_transfer_address = self
+            .client
+            .new_transfer_address(&request.name, &request.enckey)
+            .map_err(to_rpc_error)?;
+        let tx_id = self
+            .client
+            .send_to_address(
+                &request.name,
+                &request.enckey,
+                total_amount,
+                to_transfer_address,
+                vec![],
+                self.network_id,
+            )
+            .map_err(to_rpc_error)?;
+
+        // 2. waiting the transfer transaction confirmed
+        let block_height = self
+            .client
+            .get_current_block_height()
+            .map_err(to_rpc_error)?;
+        loop {
+            // query tx_id from tendermint
+            if !self.client.check_tx_confirmed(tx_id) {
+                std::thread::sleep(Duration::from_secs(1));
+                let current_block_height = self
+                    .client
+                    .get_current_block_height()
+                    .map_err(to_rpc_error)?;
+                if current_block_height - block_height >= self.block_height_ensure {
+                    return Err(rpc_error_from_string("prepare utxo failed!".into()));
+                }
+                continue;
+            }
+            break;
+        }
+
+        // 3. use the outputs[0] to deposit
+        let transaction = self
+            .client
+            .get_transaction(&request.name, &request.enckey, tx_id)
+            .map_err(to_rpc_error)?;
+        let output = match transaction {
+            Transaction::TransferTransaction(tx) => {
+                if tx.outputs.is_empty() {
+                    return Err(rpc_error_from_string("invalid transaction".into()));
+                }
+                tx.outputs[0].clone()
+            }
+            _ => return Err(rpc_error_from_string("invalid transaction type".into())),
+        };
+        let txo_pointer = TxoPointer::new(tx_id, 0);
+        let transactions = vec![(txo_pointer, output)];
         let transaction = self
             .ops_client
             .create_deposit_bonded_stake_transaction(
                 &request.name,
                 &request.enckey,
-                inputs,
-                addr,
+                transactions,
+                to_staking_address,
                 attr,
             )
             .map_err(to_rpc_error)?;
 
+        // 4. broadcast the deposit transaction and waiting it confirmed
```

Comment:
this functionality is +- duplicate with client-cli -- could this be put to client-core?

Reply:
Ok